x-pack/plugins/cases/public/containers/messages_storage.ts

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { Storage } from '../../../../../src/plugins/kibana_utils/public';
+
+let instance: MessagesStorage;
```

Comment:
Perhaps you could instead model this using a react provider and context. The provider would be used at the root of the react app and it could create an instance of MessageStorage. With that design, there would be no need for a module level binding here. The advantage is that each react app (i.e. each one created via a `React.mount` call) can create its own instance of MessagesStorage and each automated test call can create its own instance of MessagesStorage as well.

This also removes a side effect from the react app. Instead it allows the code that owns the react app to create and maintain a binding to the MessagesStorage.

Reply:
Yea, that's much cleaner and will avoid any potential cross test pollution. None of these component/views are going to be directly mounted as top level views as all the plugin is doing atm is exporting view components for other apps to consume, but def a better model and more future proof.

Reply:
Actually, thinking about it some more the module level binding may be better for now. Since each view component is consumed separately, if I were to wrap each view that uses this context in its own provider, then a single application consuming multiple case views will end up having multiple instances of storage in its subtree. We could have the consuming app provide their own instance of storage? That being said, it really won't matter as the underlying key `${plugin}-messages` will stay the same so all instances will end up acting on the same data 🤷🏾

Reply:
I see now. I didn't realize that the singleton had no state of its own. If that's the case, there is no need to make it a singleton at all. Callers can reinstate it on each use as it has no state. In fact, it doesn't need to be a class (or object) since there is no state. You can model the code as exported functions. These functions can access storage via a global binding, or you can have the functions take a reference to a storage object. If they receive a storage object, then you can test them by passing a mock storage object.

Reply:
@michaelolo24 I agree with @oatkiller. I think functions will be a better approach. I guess you didn't use `const { storage } = useKibana().services;` because some plugins will not provide that to the Kibana context, right? If that the reason you can use the same structure (`useMessagesStorage`) but instead of getting the storage from `useKibana().services` you can initiated yourself inside the hook. I don't see a problem with having multiple instance of the storage class. At the end, the browser's  `localStorage` is probably a singleton.

If you really want to have one instance of the `Storage` and not having to require it from the consumers you can pass it to your `CasesUiClient` from the `plugin.ts`. Example:

interface CasesUiClientDeps {
  storage: Storage
}

const createCasesUiClient = ({ storage }: CasesUiClientDeps) => ({
      getAllCases: (props) => {
        return getAllCasesLazy(props, { storage });
      },
      getCaseView: (props) => {
        return getCaseViewLazy(props, { storage });
      },
      getConfigureCases: (props) => {
        return getConfigureCasesLazy(props, { storage });
      },
      getCreateCase: (props) => {
        return getCreateCaseLazy(props, { storage });
      },
      getRecentCases: (props) => {
        return getRecentCasesLazy(props, { storage });
      },
      getAllCasesSelectorModal: (props) => {
        return getAllCasesSelectorModalLazy(props, { storage });
      },
})

public start(core: CoreStart, plugins: StartPlugins): CasesUiStart {
    KibanaServices.init({ ...core, ...plugins, kibanaVersion: this.kibanaVersion });
    const storage = new Storage(localStorage);
    return createCasesUiClient({ storage });
  }

Reply:
Thanks for the feedback @oatkiller and @cnasikas. After getting caught up in making these minor changes, I realized that the code that was there already is all I actually wanted 😂. This PR was initially made because we weren't sure if we would still need the singleton/service setup and then I went down a completely unnecessary rabbit hole 😂. Thanks for the feedback though! Gonna close this PR

Reply:
@cnasikas - one thing this does bring up then is that we probably don't need to be passing storage as a service in `security_solution` either..